kernel/src/process.rs

```diff
@@ -1092,14 +1132,23 @@ impl<S: 'static + UserspaceKernelBoundary> Process<'a, S> {
         (None, 0, 0)
     }
 
-    fn mem_break(&self) -> *const u8 {
-        self.kernel_memory_break.get()
-    }
-
     fn sp(&self) -> *const usize {
         self.current_stack_pointer.get() as *const usize
     }
 
+    /// Checks if the buffer represented by the passed in base pointer and size
+    /// are within the memory bounds currently exposed to the processes (i.e.
+    /// ending at `app_break`. If this method returns true, the buffer
+    /// is guaranteed to be accessible to the process and to not overlap with
+    /// the grant region.
+    fn in_app_owned_memory(&self, buf_start_addr: *const u8, size: usize) -> bool {
+        let buf_end_addr = buf_start_addr.wrapping_offset(size as isize);
+
+        buf_end_addr >= buf_start_addr
+            && buf_start_addr >= self.mem_start()
+            && buf_end_addr <= self.app_break.get()
```

Comment:
Not sure but isn't there an off-by-one here and this should be < instead of <=?

Reply:
It's fine. Imagine the app_break is 1000. That means addresses 0-999 are valid. Now suppose you have a buffer that starts at address 900 and is 100 bytes long (it is bytes 900-999). This is safe. 900 + 100 <= 1000.